network/kademlia_load_balancer.go

```diff
@@ -0,0 +1,254 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+package network
+
+import (
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network/pubsubchannel"
+	"github.com/ethersphere/swarm/network/resourceusestats"
+)
+
+// KademliaBackend is the required interface of KademliaLoadBalancer.
+type KademliaBackend interface {
+	SubscribeToPeerChanges() (addedSub *pubsubchannel.Subscription, removedPeerSub *pubsubchannel.Subscription)
+	BaseAddr() []byte
+	EachBinDesc(base []byte, minProximityOrder int, consumer PeerBinConsumer)
+	EachBinDescFiltered(base []byte, capKey string, minProximityOrder int, consumer PeerBinConsumer) error
+	EachConn(base []byte, o int, f func(*Peer, int) bool)
+}
+
+// Creates a new KademliaLoadBalancer from a KademliaBackend.
+// If useNearestNeighbourInit is true the nearest neighbour peer use count will be used when a peer is initialized.
+// If not, least used peer use count in same bin as new peer will be used. It is not clear which one is better, when
+// this load balancer would be used in several use cases we could do take some decision.
+func NewKademliaLoadBalancer(kademlia KademliaBackend, useNearestNeighbourInit bool) *KademliaLoadBalancer {
+	onPeerSub, offPeerSub := kademlia.SubscribeToPeerChanges()
+	quitC := make(chan struct{})
+	klb := &KademliaLoadBalancer{
+		kademlia:         kademlia,
+		resourceUseStats: resourceusestats.NewResourceUseStats(quitC),
+		onPeerSub:        onPeerSub,
+		offPeerSub:       offPeerSub,
+		quitC:            quitC,
+	}
+	if useNearestNeighbourInit {
+		klb.initCountFunc = klb.nearestNeighbourUseCount
+	} else {
+		klb.initCountFunc = klb.leastUsedCountInBin
+	}
+
+	go klb.listenNewPeers()
+	go klb.listenOffPeers()
+	return klb
+}
+
+// Consumer functions. A consumer is a function that uses an element returned by an iterator. It usually also returns
+// a boolean signaling if it wants to iterate more or not. We created an alias for consumer function (LBBinConsumer)
+// for code clarity.
+
+// An LBPeer represents a peer with a AddUseCount() function to signal that the peer has been used in order
+// to account it for LB sorting criteria.
+type LBPeer struct {
+	Peer  *Peer
+	stats *resourceusestats.ResourceUseStats
+}
+
+// AddUseCount is called to account a use for these peer. Should be called if the peer is actually used.
+func (lbPeer LBPeer) AddUseCount() {
+	lbPeer.stats.AddUse(lbPeer.Peer)
+}
+
+// LBBin represents a Bin of LBPeer's
+type LBBin struct {
+	LBPeers        []LBPeer
```

Comment:
Do you mean `LBPeers        []*LBPeer` ?
Slices of pointers are worst performers than regular slices, aren't? If you create a go array of structs, all memory is allocated at once, instead of n allocations. In this case at the size of the array is well known an no element is added or removed, maybe is better to have `[]LBPeer` (but maybe I'm wrong). 
And LBPeer is a very light wrapper around Peer and stats which are pointers.